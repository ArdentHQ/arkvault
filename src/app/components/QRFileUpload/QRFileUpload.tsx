import QRScanner from "qr-scanner";
import React from "react";
import { useTranslation } from "react-i18next";

import { Button } from "@/app/components/Button";
import { Icon } from "@/app/components/Icon";
import { useFiles } from "@/app/hooks/use-files";

interface QRFileUploadProperties {
	onError: (error: Error) => void;
	onRead: (text: string) => void;
}

export const QRFileUpload = ({ onError, onRead }: QRFileUploadProperties) => {
	const { t } = useTranslation();
	const { showOpenDialog } = useFiles();

	const handeQRFileScan = async () => {
		try {
			const file = await showOpenDialog({});

			if (!file) {
				return;
			}

			const { data } = await QRScanner.scanImage(file, { returnDetailedScanResult: true });

			onRead(data);
		} catch (error) {
			if (error.name === "AbortError") {
				return;
			}

			onError(new Error("InvalidQR"));
		}
	};

	return (
		<Button
			variant="secondary"
			theme="dark"
			className="z-20 space-x-2"
			onClick={handeQRFileScan}
			data-testid="QRFileUpload__upload"
		>
			<Icon name="ArrowUpBracket" />
			<span>{t("TRANSACTION.MODAL_QR_CODE.UPLOAD")}</span>
		</Button>
	);
};
