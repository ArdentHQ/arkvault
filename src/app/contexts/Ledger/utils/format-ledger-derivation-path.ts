import { LedgerDerivationScheme } from "@/app/contexts/Ledger/Ledger.contracts";

export const formatLedgerDerivationPath = (scheme: LedgerDerivationScheme) =>
	`m/${scheme.purpose || 44}'/${scheme.coinType}'/${scheme.account || 0}'/${scheme.change || 0}/${
		scheme.address || 0
	}`;
