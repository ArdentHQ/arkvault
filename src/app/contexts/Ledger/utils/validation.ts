import semver from "semver";
import { Coins } from "@payvo/sdk";
import { minVersionList } from "@/app/contexts/Ledger/Ledger.contracts";

export const hasRequiredAppVersion = async (coin: Coins.Coin) => {
	const coinId = coin.network().coin();

	if (minVersionList[coinId]) {
		const currentVersion = await coin.ledger().getVersion();

		if (semver.lt(currentVersion, minVersionList[coinId])) {
			return false;
		}
	}

	return true;
};
