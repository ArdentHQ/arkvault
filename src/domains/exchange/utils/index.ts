export * from "./assertions";
export * from "./errors";
