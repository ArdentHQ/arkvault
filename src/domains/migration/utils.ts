import { MigrationTransactionStatus } from "@/domains/migration/migration.contracts";

export const getIcon = (status: MigrationTransactionStatus) => {
	if (status === MigrationTransactionStatus.Confirmed) {
		return {
			color: "text-theme-success-600",
			name: "CircleCheckMark",
		};
	}

	return {
		color: "text-theme-warning-300",
		name: "Clock",
	};
};
