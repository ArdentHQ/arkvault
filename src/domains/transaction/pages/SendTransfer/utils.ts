export const getFeeType = (numberOfRecipients: number): "multiPayment" | "transfer" =>
	numberOfRecipients > 1 ? "multiPayment" : "transfer";
