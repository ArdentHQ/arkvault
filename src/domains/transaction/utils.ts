import { Services } from "@ardenthq/sdk";
import { RecipientItem } from "@/domains/transaction/components/RecipientList/RecipientList.contracts";

export const isNoDeviceError = (error: any) => {
	if (!error) {
		return false;
	}

	if (String(error).includes("no device found")) {
		return true;
	}

	return false;
};

export const isRejectionError = (error: any) => String(error).includes("Condition of use not satisfied");

export const handleBroadcastError = ({ errors }: Services.BroadcastResponse) => {
	const allErrors = Object.values(errors);

	if (allErrors.length === 0) {
		return;
	}

	throw new Error(allErrors[0]);
};

export const getTransferType = ({ recipients }: { recipients: RecipientItem[] }): "multiPayment" | "transfer" => {
	const isMultiPayment = recipients.length > 1;
	return isMultiPayment ? "multiPayment" : "transfer";
};

export const withAbortPromise =
	(signal?: AbortSignal, callback?: () => void) =>
	<T>(promise: Promise<T>) =>
		new Promise<T>((resolve, reject) => {
			if (signal) {
				signal.addEventListener("abort", () => {
					callback?.();
					reject("ERR_ABORT");
				});
			}

			return promise.then(resolve).catch(reject);
		});
