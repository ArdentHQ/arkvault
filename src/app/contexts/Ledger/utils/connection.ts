import { Coins } from "@ardenthq/sdk";
import retry, { AbortError, Options } from "p-retry";

import { formatLedgerDerivationPath } from "./format-ledger-derivation-path";
import { hasRequiredAppVersion } from "./validation";

export const accessLedgerDevice = async (coin: Coins.Coin) => {
	try {
		await coin.__construct();
		await coin.ledger().connect();
	} catch (error) {
		// If the device is open, continue normally.
		// Can be triggered when the user retries ledger connection.
		if (error.message !== "The device is already open.") {
			throw error;
		}
	}
};

export const accessLedgerApp = async ({ coin }: { coin: Coins.Coin }) => {
	await accessLedgerDevice(coin);

	if (!(await hasRequiredAppVersion(coin))) {
		throw new Error("VERSION_ERROR");
	}

	// Ensure that the app is accessible.
	await coin.ledger().getPublicKey(
		formatLedgerDerivationPath({
			coinType: coin.config().get<number>("network.constants.slip44"),
		}),
	);
};

export const persistLedgerConnection = async ({
	coin,
	options,
	hasRequestedAbort,
}: {
	coin: Coins.Coin;
	options: Options;
	hasRequestedAbort: () => boolean;
}) => {
	const retryAccess: any = async (attempts: number) => {
		if (hasRequestedAbort() && attempts > 1) {
			throw new AbortError("CONNECTION_ERROR");
		}

		try {
			await accessLedgerApp({ coin });
		} catch (error) {
			// Abort on version error or continue retrying access.
			if (error.message === "VERSION_ERROR") {
				throw new AbortError("VERSION_ERROR");
			}

			throw error;
		}
	};

	await retry(retryAccess, options);
};
