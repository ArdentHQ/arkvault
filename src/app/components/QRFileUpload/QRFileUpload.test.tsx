import userEvent from "@testing-library/user-event";
import * as browserAccess from "browser-fs-access";
import QRScanner from "qr-scanner";
import React from "react";

import { render, screen, waitFor } from "@/utils/testing-library";

import { QRFileUpload } from "./QRFileUpload";

const qrCodeUrl =
	"http://localhost:3000/#/?amount=10&coin=ARK&method=transfer&network=ark.devnet&recipient=DNSBvFTJtQpS4hJfLerEjSXDrBT7K6HL2o";

describe("QRFileUpload", () => {
	it("should render", () => {
		const { asFragment } = render(<QRFileUpload />);

		expect(asFragment()).toMatchSnapshot();
	});

	it("should read qr code", async () => {
		const onRead = vi.fn();
		const browserAccessMock = vi.spyOn(browserAccess, "fileOpen").mockResolvedValue(new File([], "test.png"));

		const scanImageMock = vi.spyOn(QRScanner, "scanImage").mockResolvedValue({ data: qrCodeUrl });

		render(<QRFileUpload onError={vi.fn()} onRead={onRead} />);

		await userEvent.click(screen.getByTestId("QRFileUpload__upload"));
		await waitFor(() => expect(onRead).toHaveBeenCalledWith(qrCodeUrl));

		browserAccessMock.mockRestore();
		scanImageMock.mockRestore();
	});

	it("should stay idle if qr code file is not selected from file dialog", async () => {
		const onRead = vi.fn();
		const browserAccessMock = vi.spyOn(browserAccess, "fileOpen").mockResolvedValue(undefined);

		const scanImageMock = vi.spyOn(QRScanner, "scanImage").mockResolvedValue({ data: qrCodeUrl });

		render(<QRFileUpload onError={vi.fn()} onRead={onRead} />);

		await userEvent.click(screen.getByTestId("QRFileUpload__upload"));
		await waitFor(() => expect(onRead).not.toHaveBeenCalled());

		browserAccessMock.mockRestore();
		scanImageMock.mockRestore();
	});

	it("should stay idle if use closed file dialog", async () => {
		const onRead = vi.fn();
		const onError = vi.fn();
		const browserAccessMock = vi.spyOn(browserAccess, "fileOpen").mockImplementation(() => {
			const error = new Error("AbortError");
			error.name = "AbortError";

			throw error;
		});

		const scanImageMock = vi.spyOn(QRScanner, "scanImage").mockResolvedValue({ data: qrCodeUrl });
		render(<QRFileUpload onRead={onRead} onError={onError} />);

		await userEvent.click(screen.getByTestId("QRFileUpload__upload"));

		await waitFor(() => expect(onRead).not.toHaveBeenCalled());
		await waitFor(() => expect(onError).not.toHaveBeenCalled());

		browserAccessMock.mockRestore();
		scanImageMock.mockRestore();
	});

	it("should emit error if qr code read fails", async () => {
		const onError = vi.fn();
		const browserAccessMock = vi.spyOn(browserAccess, "fileOpen").mockResolvedValue(new File([], "test.png"));
		const errorMessage = "InvalidQR";

		const scanImageMock = vi.spyOn(QRScanner, "scanImage").mockImplementation(() => {
			throw new Error(errorMessage);
		});

		render(<QRFileUpload onError={onError} onRead={vi.fn()} />);

		await userEvent.click(screen.getByTestId("QRFileUpload__upload"));
		await waitFor(() => expect(onError).toHaveBeenCalledWith(new Error(errorMessage)));

		browserAccessMock.mockRestore();
		scanImageMock.mockRestore();
	});
});
