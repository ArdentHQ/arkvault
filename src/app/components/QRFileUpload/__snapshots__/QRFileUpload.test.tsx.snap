// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`QRFileUpload > should render 1`] = `
<DocumentFragment>
  <button
    class="px-5 py-3 text-base relative items-center inline-flex justify-center font-semibold leading-tight text-center transition-colors-shadow duration-100 ease-linear outline-none rounded focus:outline-none focus:ring-2 focus:ring-theme-primary-400 disabled:cursor-not-allowed dark:bg-theme-secondary-800 dark:text-theme-secondary-200 bg-theme-primary-100 text-theme-primary-600 hover:bg-theme-primary-700 hover:text-white dark:hover:bg-theme-primary-700 z-20 space-x-2 sm:bg-theme-secondary-800 sm:text-theme-secondary-200"
    data-testid="QRFileUpload__upload"
    type="button"
  >
    <div
      class="flex items-center space-x-2"
    >
      <div
        class="css-v0ob3f"
        height="16"
        width="16"
      >
        <svg
          id="arrow-up-bracket"
          viewBox="0 0 20 20"
          x="0"
          xml:space="preserve"
          xmlns="http://www.w3.org/2000/svg"
          y="0"
        >
          <path
            d="M19 13.4v1.9c0 2.1-1.7 3.7-3.7 3.7H4.7c-2 0-3.7-1.7-3.7-3.7v-1.9m9.05-.3V2m-4.7 4.7l4.7-4.7 4.6 4.7"
            fill="none"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </div>
      <span>
        Upload Image
      </span>
    </div>
  </button>
</DocumentFragment>
`;
